defi/lending_platform/src/lending_platform.rs

```diff
@@ -0,0 +1,436 @@
+            let users_loan_balance = self.calculate_total_loan_balance(user);
+            let available_collateral = users_total_collateral - users_loan_balance;
+            info!(
+                "[LendingPlatform][USER:{}] Total collateral of {} XRD. Current loan balance of \
+                {} XRD. Available collateral of {} XRD.",
+                user_address,
+                users_total_collateral,
+                users_loan_balance,
+                available_collateral
+            );
+            available_collateral
+        }
+
+        pub fn new() -> (Component, Bucket) {
+            let admin_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                .metadata("name", "Lending Platform Admin Badge")
+                .initial_supply_fungible(1);
+
+            let lending_platform = Self {
+                assets: LazyMap::new(),
+                asset_ltv_ratios: LazyMap::new(),
+                loan_balances: LazyMap::new(),
+                users: LazyMap::new(),
+                admin_badge: admin_badge.resource_def()
+            }
+            .instantiate();
+            (lending_platform, admin_badge)
+        }
+
+        pub fn new_user(&self) -> Bucket { // return a badge
+            let user_badge = ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                .metadata("name", "Lending Platform Badge")
+                .initial_supply_fungible(1);
+
+            let user = User {
+                    user_address: user_badge.resource_address(),
+                    deposit_balances: HashMap::new(),
+                    borrow_balances: HashMap::new()
+                };
+            self.users.insert(user_badge.resource_address(), user);
+            info!(
+                "[LendingPlatform][POOL] Created new user with an address of {}.",
+                user_badge.resource_address()
+            );
+            user_badge
+        }
+
+        #[auth(admin_badge)]
+        pub fn new_asset(&self, asset_address: Address, ltv_ratio: Decimal) {
+            assert! (
+                ltv_ratio >= 0.into() && ltv_ratio <= 1.into(),
+                "[LendingPlatform][POOL] LTV Ratio must be between 0.0 & 1.0 (inclusive)."
+                );
+            match self.asset_ltv_ratios.get(&asset_address) {
+                Some(..) =>  {
+                    panic!("[LendingPlatform][POOL] Asset {} already exists in lending pool.", asset_address);
+                },
+                None => {
+                    self.asset_ltv_ratios.insert(asset_address, ltv_ratio);
+                    info!(
+                        "[LendingPlatform][POOL] Added asset {} with LTV ratio of {} to the lending pool.",
+                        asset_address,
+                        ltv_ratio
+                    )
+                }
+            };
+        }
+
+        pub fn deposit_asset(&self, user_auth: BucketRef, asset: Bucket) -> BucketRef {
```

Comment:
No need to return the BucketRef here. You can just call `user_auth.drop()` somewhere inside the method. Or better: Make the user auth a NFT, use the `#[auth(user_auth_nft)]` macro which will automatically drop the reference

Reply:
Updated it to use `user_auth.drop()`.

To use the `#[auth(user_auth_nft)]` macro, would I need to store information about the badge within the component similar to how I did the `admin_badge` auth? Or is there an alternative way to do this that doesn't require having access to `user_auth_nft` from within the component?

Reply:
You would do exactly the same thing as you are doing right now but instead of using the `user_auth.resource_address` as keys to the maps, you would use the NFT id

Reply:
What would the function signature look like? If I do
```
#[auth(user_auth)]
pub fn deposit_asset(&self, asset: Bucket)
```

I get the following error because the nft does not belong to the component
```
error[E0425]: cannot find value `user_auth` in this scope
   --> src/lending_platform.rs:167:44
    |
167 |             let mut user = self.users.get(&user_auth.resource_address()).unwrap();
    |                                            ^^^^^^^^^ not found in this scope

error[E0425]: cannot find value `user_auth` in this scope
   --> src/lending_platform.rs:180:31
    |
180 |             self.users.insert(user_auth.resource_address(), user);
    |                               ^^^^^^^^^ not found in this scope

error[E0609]: no field `user_auth` on type `&lending_platform::blueprint::LendingPlatform`
   --> src/lending_platform.rs:162:16
    |
162 |         #[auth(user_auth)]
    |                ^^^^^^^^^ unknown field
    |
    = note: available fields are: `assets`, `asset_ltv_ratios`, `loan_balances`, `users`, `admin_badge`
```

Not sure how to achieve this without making the badges part of the struct

Reply:
To reference the resource used as authentication, you use the `auth` variable. In that case, you used the name of the variable (`user_auth`)

Reply:
Let's say you name the NFT resourceDef `auth_nft` in you component's state:
```
struct LendingPlatform {
   assets: LazyMap<Address, Vault>,
   asset_ltv_ratios: LazyMap<Address, Decimal>,
   ...
   auth_nft: ResourceDef 
}
```

You would reference it like this in the auth macro:

```
#[auth(auth_nft)]
pub fn deposit_asset(&self, asset: Bucket)
```

The problem might be that you are not saving the NFT resource def in the state ?